File: solutions/90. Subsets II.java
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        List<List<Integer>> subsets=new ArrayList<>();
        List<Integer> set=new ArrayList<>();
        Arrays.sort(nums);
        func(nums,0,set,subsets);
        return subsets;
    }
​
    private void func(int[] nums, int i, List<Integer> set, List<List<Integer>> subsets) {
        subsets.add(new ArrayList<>(set));
        for (int index=i; index< nums.length;index++){
           if (i!=index && nums[index] == nums[index-1]) continue;
           set.add(nums[index]);
           func(nums,index+1,set, subsets);
           set.remove(set.size()-1);
        }
    }
}

File: ..solutions/90. Subsets II.cpp
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        /*
            * iterative approch exists, but need O(N) extra space
            * n = size of vector<int>nums.
        */
        sort(nums.begin(), nums.end());
        vector<vector<int>> subsets;
        dfs(nums, subsets, 0, vector<int>());
        return subsets;
    }
    
    void dfs(vector<int> &nums, vector<vector<int>> &subsets, int index, vector<int> current){
        subsets.push_back(current);
        for(int i = index; i < (int) nums.size(); i++){
            if(index < i && nums[i-1] == nums[i]){
                continue;
            }
            current.push_back(nums[i]);
            dfs(nums, subsets, i+1, current);
            current.pop_back();
        }
    }
};
